File: src/main/java/com/alxnns1/mobhunter/init/MHBlocks.java
package com.alxnns1.mobhunter.init;

import com.alxnns1.mobhunter.blocks.MHPlant;
import net.minecraft.block.*;
import net.minecraft.block.material.Material;
import net.minecraft.client.renderer.RenderType;
import net.minecraft.client.renderer.RenderTypeLookup;
import net.minecraftforge.client.event.ColorHandlerEvent;
import net.minecraftforge.common.ToolType;

import java.util.HashMap;

public class MHBlocks {

	public static final HashMap<String, Block> BLOCKS = new HashMap<>();
	private static final HashMap<Block, Integer> PLANT_BLOCKS = new HashMap<>();

	public static Block[] register() {
		//HERBS
		registerPlantBlock("herb", MHColours.GREEN);
		registerPlantBlock("antidote_herb", MHColours.BLUE);
		registerPlantBlock("fire_herb", MHColours.RED);
		registerPlantBlock("ivy", MHColours.GREEN);
		registerPlantBlock("sleep_herb", MHColours.LIGHT_BLUE);
		registerPlantBlock("sap_plant", MHColours.WHITE);
		registerPlantBlock("felvine", MHColours.YELLOW);
		registerPlantBlock("gloamgrass_root", MHColours.PINK);
		registerPlantBlock("hot_pepper", MHColours.RED);
		registerPlantBlock("snow_herb", MHColours.LIGHT_BLUE);
		registerPlantBlock("dragonvine", MHColours.PURPLE);
		registerPlantBlock("tough_dragonvine", MHColours.PURPLE);
		//MUSHROOMS
		registerPlantBlock("blue_mushroom", MHColours.BLUE);
		registerPlantBlock("nitroshroom", MHColours.ORANGE);
		registerPlantBlock("parashroom", MHColours.YELLOW);
		registerPlantBlock("toadstool", MHColours.PURPLE);
		registerPlantBlock("mopeshroom", MHColours.LIGHT_BLUE);
		registerPlantBlock("exciteshroom", MHColours.GREEN);
		registerPlantBlock("dragon_toadstool", MHColours.RED);
		//BERRIES
		registerPlantBlock("paintberry", MHColours.PINK);
		registerPlantBlock("might_seed", MHColours.RED);
		registerPlantBlock("adamant_seed", MHColours.ORANGE);
		registerPlantBlock("nulberry", MHColours.BLUE);
		registerPlantBlock("dragonfell_berry", MHColours.RED);
		registerPlantBlock("scatternut", MHColours.GREY);
		registerPlantBlock("needleberry", MHColours.GREY);
		registerPlantBlock("latchberry", MHColours.WHITE);
		registerPlantBlock("bomberry", MHColours.GREY);
		//ORES
		registerOreBlock("earth_crystal_ore", properties(Material.ROCK, 1));
		registerOreBlock("machalite_ore", properties(Material.ROCK, 1));
		registerOreBlock("ice_crystal_ore", properties(Material.ROCK, 1));
		registerOreBlock("bealite_ore", properties(Material.ROCK, 1));
		registerOreBlock("lightcrystal_ore", properties(Material.ROCK, 1).setLightLevel(state -> 7));
		registerOreBlock("dragonite_ore", properties(Material.ROCK, 1));
		registerOreBlock("firestone_ore", properties(Material.ROCK, 1));
		registerOreBlock("carbalite_ore", properties(Material.ROCK, 2));
		registerOreBlock("eltalite_ore", properties(Material.ROCK, 3));
		registerOreBlock("meldspar_ore", properties(Material.ROCK, 3));
		registerOreBlock("novacrystal_ore", properties(Material.ROCK, 2).setLightLevel(state -> 14));
		registerOreBlock("purecrystal_ore", properties(Material.ROCK, 3));
		registerOreBlock("fucium_ore", properties(Material.ROCK, 2));
		registerOreBlock("firecell_stone_ore", properties(Material.ROCK, 2));
		registerOreBlock("allfire_stone_ore", properties(Material.ROCK, 3));
		registerOreBlock("ultimas_crystal_ore", properties(Material.ROCK, 3));
		//COMPACT BLOCKS
		registerBlock("earth_crystal_block", properties(Material.IRON, 1));
		registerBlock("machalite_block", properties(Material.IRON, 1));
		registerBlock("ice_crystal_block", properties(Material.IRON, 1));
		registerBlock("bealite_block", properties(Material.IRON, 1));
		registerBlock("lightcrystal_block", properties(Material.IRON, 1).setLightLevel(state -> 7));
		registerBlock("dragonite_block", properties(Material.IRON, 1));
		registerBlock("firestone_block", properties(Material.IRON, 1));
		registerBlock("carbalite_block", properties(Material.IRON, 2));
		registerBlock("eltalite_block", properties(Material.IRON, 3));
		registerBlock("meldspar_block", properties(Material.IRON, 3));
		registerBlock("novacrystal_block", properties(Material.IRON, 2).setLightLevel(state -> 14));
		registerBlock("purecrystal_block", properties(Material.IRON, 3));
		registerBlock("fucium_block", properties(Material.IRON, 2));
		registerBlock("firecell_stone_block", properties(Material.IRON, 2));
		registerBlock("allfire_stone_block", properties(Material.IRON, 3));
		registerBlock("ultimas_crystal_block", properties(Material.IRON, 3));

		return BLOCKS.values().toArray(new Block[0]);
	}

	public static void setRenderLayerForPlants() {
		PLANT_BLOCKS.keySet().forEach(block -> RenderTypeLookup.setRenderLayer(block, RenderType.getCutoutMipped()));
	}

	public static void setColourForPlants(ColorHandlerEvent.Block event) {
		PLANT_BLOCKS.forEach((block, colour) -> event.getBlockColors().register((blockStateIn, lightReaderIn, blockPosIn, tintIndexIn) -> colour, block));
	}

	private static AbstractBlock.Properties properties(Material material, int harvestLevel) {
		return Block.Properties.create(material).setRequiresTool().harvestTool(ToolType.PICKAXE).harvestLevel(harvestLevel).hardnessAndResistance(3F, 3F);
	}

	private static void registerBlock(String name, Block.Properties properties) {
		Block block = new Block(properties).setRegistryName(name);
		BLOCKS.put(name, block);
	}

	private static void registerOreBlock(String name, Block.Properties properties) {
		Block block = new OreBlock(properties).setRegistryName(name);
		BLOCKS.put(name, block);
	}

	private static void registerPlantBlock(String name, int colour) {
		Block.Properties properties = Block.Properties.create(Material.PLANTS).doesNotBlockMovement().zeroHardnessAndResistance().sound(SoundType.PLANT);
		Block block = new MHPlant(properties).setRegistryName(name);
		BLOCKS.put(name, block);
		PLANT_BLOCKS.put(block, colour);
	}
}

File: src/main/kotlin/com/alxnns1/mobhunter/init/MHBlocks.kt
package com.alxnns1.mobhunter.init

import com.alxnns1.mobhunter.block.MHPlant
import net.minecraft.block.AbstractBlock.Properties
import net.minecraft.block.Block
import net.minecraft.block.OreBlock
import net.minecraft.block.SoundType
import net.minecraft.block.material.Material
import net.minecraft.client.renderer.RenderType
import net.minecraft.client.renderer.RenderTypeLookup
import net.minecraftforge.api.distmarker.Dist
import net.minecraftforge.api.distmarker.OnlyIn
import net.minecraftforge.client.event.ColorHandlerEvent
import net.minecraftforge.common.ToolType
import net.minecraftforge.event.RegistryEvent
import thedarkcolour.kotlinforforge.forge.objectHolder

object MHBlocks {
	val BLOCKS = mutableListOf<Block>()
	private val PLANTS = mutableListOf<MHPlant>()

	// Herbs
	val HERB: Block by objectHolder("herb")
	val ANTIDOTE_HERB: Block by objectHolder("antidote_herb")
	val FIRE_HERB: Block by objectHolder("fire_herb")
	val IVY: Block by objectHolder("ivy")
	val SLEEP_HERB: Block by objectHolder("sleep_herb")
	val SAP_PLANT: Block by objectHolder("sap_plant")
	val FELVINE: Block by objectHolder("felvine")
	val GLOAMGRASS_ROOT: Block by objectHolder("gloamgrass_root")
	val HOT_PEPPER: Block by objectHolder("hot_pepper")
	val SNOW_HERB: Block by objectHolder("snow_herb")
	val DRAGONVINE: Block by objectHolder("dragonvine")
	val TOUGH_DRAGONVINE: Block by objectHolder("tough_dragonvine")

	// Mushrooms
	val BLUE_MUSHROOM: Block by objectHolder("blue_mushroom")
	val NITROSHROOM: Block by objectHolder("nitroshroom")
	val PARASHROOM: Block by objectHolder("parashroom")
	val TOADSTOOL: Block by objectHolder("toadstool")
	val MOPESHROOM: Block by objectHolder("mopeshroom")
	val EXCITESHROOM: Block by objectHolder("exciteshroom")
	val DRAGON_TOADSTOOL: Block by objectHolder("dragon_toadstool")

	// Berries
	val PAINTBERRY: Block by objectHolder("paintberry")
	val MIGHT_SEED: Block by objectHolder("might_seed")
	val ADAMANT_SEED: Block by objectHolder("adamant_seed")
	val NULBERRY: Block by objectHolder("nulberry")
	val DRAGONFELL_BERRY: Block by objectHolder("dragonfell_berry")
	val SCATTERNUT: Block by objectHolder("scatternut")
	val NEEDLEBERRY: Block by objectHolder("needleberry")
	val LATCHBERRY: Block by objectHolder("latchberry")
	val BOMBERRY: Block by objectHolder("bomberry")

	// Ores
	val EARTH_CRYSTAL_ORE: Block by objectHolder("earth_crystal_ore")
	val MACHALITE_ORE: Block by objectHolder("machalite_ore")
	val ICE_CRYSTAL_ORE: Block by objectHolder("ice_crystal_ore")
	val BEALITE_ORE: Block by objectHolder("bealite_ore")
	val LIGHTCRYSTAL_ORE: Block by objectHolder("lightcrystal_ore")
	val DRAGONITE_ORE: Block by objectHolder("dragonite_ore")
	val FIRESTONE_ORE: Block by objectHolder("firestone_ore")
	val ALLFIRE_STONE_ORE: Block by objectHolder("allfire_stone_ore")
	val CARBALITE_ORE: Block by objectHolder("carbalite_ore")
	val ELTALITE_ORE: Block by objectHolder("eltalite_ore")
	val MELDSPAR_ORE: Block by objectHolder("meldspar_ore")
	val NOVACRYSTAL_ORE: Block by objectHolder("novacrystal_ore")
	val PURECRYSTAL_ORE: Block by objectHolder("purecrystal_ore")
	val FUCIUM_ORE: Block by objectHolder("fucium_ore")
	val FIRECELL_STONE_ORE: Block by objectHolder("firecell_stone_ore")
	val ULTIMAS_CRYSTAL_ORE: Block by objectHolder("ultimas_crystal_ore")

	// Compact Blocks
	val MACHALITE_BLOCK: Block by objectHolder("machalite_block")

	fun register(event: RegistryEvent.Register<Block>) {
		event.registry.registerAll(
			// Herbs
			plant("herb", GREEN),
			plant("antidote_herb", BLUE),
			plant("fire_herb", RED),
			plant("ivy", GREEN),
			plant("sleep_herb", CYAN),
			plant("sap_plant", WHITE),
			plant("felvine", YELLOW),
			plant("gloamgrass_root", PINK),
			plant("hot_pepper", RED),
			plant("snow_herb", BLUE),
			plant("dragonvine", PURPLE),
			plant("tough_dragonvine", PURPLE),
			// Mushrooms
			plant("blue_mushroom", BLUE),
			plant("nitroshroom", RED),
			plant("parashroom", YELLOW),
			plant("toadstool", PURPLE),
			plant("mopeshroom", CYAN),
			plant("exciteshroom", PURPLE),
			plant("dragon_toadstool", RED),
			plant("unique_mushroom", WHITE),
			plant("choice_mushroom", WHITE),
			// Berries
			plant("paintberry", PINK),
			plant("might_seed", RED),
			plant("adamant_seed", ORANGE),
			plant("nulberry", BLUE),
			plant("dragonfell_berry", RED),
			plant("scatternut", GREY),
			plant("needleberry", GREY),
			plant("latchberry", CYAN),
			plant("bomberry", GREY),
			// Ores
			oreBlock("earth_crystal_ore"),
			oreBlock("machalite_ore"),
			oreBlock("ice_crystal_ore"),
			oreBlock("bealite_ore"),
			oreBlock("lightcrystal_ore", rockProps().setLightLevel { 7 }),
			oreBlock("dragonite_ore"),
			oreBlock("firestone_ore"),
			oreBlock("allfire_stone_ore", rockProps(harvestLevel = 3)),
			oreBlock("carbalite_ore", rockProps(harvestLevel = 2)),
			oreBlock("eltalite_ore", rockProps(harvestLevel = 3)),
			oreBlock("meldspar_ore", rockProps(harvestLevel = 3)),
			oreBlock("novacrystal_ore", rockProps(harvestLevel = 2).setLightLevel { 14 }),
			oreBlock("purecrystal_ore", rockProps(harvestLevel = 3)),
			oreBlock("fucium_ore", rockProps(harvestLevel = 2)),
			oreBlock("firecell_stone_ore", rockProps(harvestLevel = 2)),
			oreBlock("ultimas_crystal_ore", rockProps(harvestLevel = 3)),
			// Compact Blocks
			block("earth_crystal_block", rockProps(material = Material.IRON)),
			block("machalite_block", rockProps(material = Material.IRON)),
			block("ice_crystal_block", rockProps(material = Material.IRON)),
			block("bealite_block", rockProps(material = Material.IRON)),
			block("lightcrystal_block", rockProps(material = Material.IRON).setLightLevel { 7 }),
			block("dragonite_block", rockProps(material = Material.IRON)),
			block("firestone_block", rockProps(material = Material.IRON)),
			block("allfire_stone_block", rockProps(material = Material.IRON, harvestLevel = 3)),
			block("carbalite_block", rockProps(material = Material.IRON, harvestLevel = 2)),
			block("eltalite_block", rockProps(material = Material.IRON, harvestLevel = 3)),
			block("meldspar_block", rockProps(material = Material.IRON, harvestLevel = 3)),
			block("novacrystal_block", rockProps(material = Material.IRON, harvestLevel = 2).setLightLevel { 14 }),
			block("purecrystal_block", rockProps(material = Material.IRON, harvestLevel = 3)),
			block("fucium_block", rockProps(material = Material.IRON, harvestLevel = 2)),
			block("firecell_stone_block", rockProps(material = Material.IRON, harvestLevel = 2)),
			block("ultimas_crystal_block", rockProps(material = Material.IRON, harvestLevel = 3))
		)
	}

	@OnlyIn(Dist.CLIENT)
	fun setRenderLayers(): Unit = PLANTS.forEach { RenderTypeLookup.setRenderLayer(it, RenderType.getCutoutMipped()) }

	@OnlyIn(Dist.CLIENT)
	fun registerBlockColours(event: ColorHandlerEvent.Block) {
		event.blockColors.register(
			{ state, _, _, _ -> state.block.let { if (it is MHPlant) it.colour else -1 } },
			*PLANTS.toTypedArray()
		)
	}

	private fun plantProps(): Properties = Properties.create(Material.PLANTS)
		.doesNotBlockMovement()
		.zeroHardnessAndResistance()
		.sound(SoundType.PLANT)

	private fun rockProps(
		material: Material = Material.ROCK,
		harvestLevel: Int = 1,
		hardness: Float = 3F,
		resistance: Float = 3F
	): Properties = Properties.create(material)
		.setRequiresTool()
		.harvestTool(ToolType.PICKAXE)
		.harvestLevel(harvestLevel)
		.hardnessAndResistance(hardness, resistance)

	private fun plant(name: String, colour: Int, props: Properties = plantProps()): Block =
		MHPlant(colour, props).also {
			it.setRegistryName(name)
			BLOCKS += it
			PLANTS += it
		}

	private fun block(name: String, props: Properties = rockProps()): Block = Block(props).also {
		it.setRegistryName(name)
		BLOCKS += it
	}

	private fun oreBlock(name: String, props: Properties = rockProps()): Block = OreBlock(props).also {
		it.setRegistryName(name)
		BLOCKS += it
	}
}
